Fix: Chat - Link in end of line displays tooltip over text and not on link

ewanmellor opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Add a new prop to `Tooltip` called `shouldUseMultilinePositioning`. This enables a new algorithm for finding the bounding box target for the tooltip. In the case where the target has wrapped onto multiple lines (e.g. it is a link in a chat window) then the link has multiple bounding boxes, and we want to show the tooltip against the one that the user is hovering over.

As part of this, extend `Hoverable` to pass the Event to its `onHoverIn` / `onHoverOut` callbacks.

Enable this new algorithm from `BaseAnchorForCommentsOnly`, which is the base class for links that show up in chat.

### Fixed Issues

<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/27585
PROPOSAL: https://github.com/Expensify/App/issues/27585#issuecomment-1724380334


### Tests

Action Performed:

On a platform with a mouse (i.e. not mobile):

1. Open the app
2. Open any report
3. Type in some text and at the end of line, add link with long text eg: Test test test test test test test test test test
4. Send the text, text should have text and link in first line and link should be continued in second line too. Adjust the window size to force the text to wrap if necessary.
5. Hover on link and observe that tooltip is displayed over text

Expected Result:
App should display tooltip over link on hover. The tooltip should appear over the same section of the link that the user is hovering over.

- [ ] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

N/A.  There is no network interaction in this change.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<img width="561" alt="Screenshot 2023-09-19 at 3 38 28 PM" src="https://github.com/Expensify/App/assets/894011/5671a313-84fd-4bf2-9806-a1ae25d5b39d">
<img width="516" alt="Screenshot 2023-09-19 at 3 38 23 PM" src="https://github.com/Expensify/App/assets/894011/82a66d2b-bbc8-47c0-8ff2-bfef3c14f0f5">

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

ewanmellor committed:
Fix: Chat - Link in end of line displays tooltip over text and not on link

Add a new prop to Tooltip called shouldUseMultilinePositioning. This
enables a new algorithm for finding the bounding box target for the
tooltip. In the case where the target has wrapped onto multiple lines
(e.g. it is a link in a chat window) then the link has multiple bounding
boxes, and we want to show the tooltip against the one that the user is
hovering over.

As part of this, extend Hoverable to pass the Event to its onHoverIn /
onHoverOut callbacks.

Enable this new algorithm from BaseAnchorForCommentsOnly, which is the
base class for links that show up in chat.

ewanmellor commented:
I have read the CLA Document and I hereby sign the CLA

ewanmellor commented:
recheck

ewanmellor commented:
recheck

s77rt commented on src/components/Tooltip/tooltipPropTypes.js:
```diff
@@ -42,6 +59,7 @@ const defaultProps = {
     renderTooltipContent: undefined,
     renderTooltipContentKey: [],
     shouldHandleScroll: false,
+    shouldUseMultilinePositioning: false,
```
Let's remove this prop. I think we want to have this feature enabled for all tooltips

s77rt commented on src/components/Hoverable/index.js:
```diff
@@ -94,7 +95,7 @@ class Hoverable extends Component {
 
         if (isHovered !== this.state.isHovered) {
-            this.setState({isHovered}, isHovered ? this.props.onHoverIn : this.props.onHoverOut);
+            this.setState({isHovered}, () => (isHovered ? this.props.onHoverIn : this.props.onHoverOut)(ev));
```
Given that `onHoverIn` may be called without events it's better to use `onMouseEnter`. Add onMouseEnter and onMouseLeave props to Hoverable and use onMouseEnter to set the target and mouse position.

In Tooltip:
```js
    const target = useRef(null);
    const initialMousePosition = useRef({x: 0, y: 0});

    const updateTargetAndMousePosition = useCallback((e) => {
        target.current = e.target;
        initialMousePosition.current = {x: e.clientX, y: e.clientY};
    }, []);

    <Hoverable onMouseEnter={updateTargetAndMousePosition) ... />

```

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -44,33 +79,59 @@ function Tooltip(props) {
+     */
+    const showTooltip = useCallback(
+        (ev) => {
+            if (shouldUseMultilinePositioning) {
+                if (ev) {
+                    const {clientX, clientY, target} = ev;
+                    const bb = chooseBoundingBox(target, clientX, clientY);
+                    updateBounds(bb);
+                }
+            }
```
Undo this. `chooseBoundingBox` should be called from `updateBounds`


```js
    const updateBounds = (bounds) => {
        if (bounds.width === 0) {
            setIsRendered(false);
        }
        const betterBounds = chooseBoundingBox(target.current, initialMousePosition.current.x, initialMousePosition.current.y);
        setWrapperWidth(betterBounds.width);
        setWrapperHeight(betterBounds.height);
        setXOffset(betterBounds.x);
        setYOffset(betterBounds.y);
    };

```

ewanmellor replied:
@s77rt This doesn't make sense. If the user hovers over one part of the link and then moves to the other part, you need to move the tooltip. `updateBounds` is only called when the element moves, but in that case the element itself has not moved, so we will miss the necessary update. That's why it needs to be in the hover event handler.

s77rt replied:
I'm not sure if this is a valid case. if the user moves to the other part then we'd have two possible outcomes:
1. User is still hovering the target, in this case the tooltip does not need to be updated as this case is only possible if the mouse stays at the same row (x position)
2. User lost hover, in this case onMouseEnter will be called once the user hover the next part and the tooltip will be placed correctly

ewanmellor replied:
@s77rt Right, but your proposed `updateTargetAndMousePosition` doesn't set the tooltip position, it just saves the mouse position at that time. So when the `onMouseEnter` is called, you remember the location but don't do anything with that. And you won't get a call to `updateBounds` because the element hasn't moved.

s77rt replied:
Yes, that's the intended use. updateBounds will be called once the tooltip is visible or on bounds change

ewanmellor replied:
@s77rt I think we're talking at cross-purposes.

Earlier you said "Undo this. chooseBoundingBox should be called from updateBounds". I took this to mean "don't call chooseBoundingBox from showTooltip, only call it from updateBounds". So how does "updateBounds will be called once the tooltip is visible" happen in your mind?

s77rt replied:
updateBounds will be called by BoundsObserver

ewanmellor replied:
@s77rt "updateBounds will be called by BoundsObserver". This only happens if the element moves. If the user is moving the mouse between two parts of the same link, the element does not move and updateBounds is not called.

s77rt replied:
That's correct. But the user can't move the mouse between two parts without firing the onMouseLeave and then onMouseEnter again on the second part

ewanmellor replied:
@s77rt Yes, but again, if the user does onMouseLeave and onMouseEnter, then you have nothing calling `updateBounds`. Your proposed `updateTargetAndMousePosition` does not call `updateBounds`. So the tooltip does not move. What am I missing from what you are saying?

s77rt replied:
- onMouseLeave -> onHoverOut -> hideTooltip -> setIsVisible(false)
- onMouseEnter (1st callback) -> updateTargetAndMousePosition
- onMouseEnter (2nd callback) -> onHoverIn -> showTooltip -> setIsVisible(true)

Besides setting the target and mouse position, we have a change in the isVisible state which is passed to the enabled prop of BoundsObserver and per the docs https://socket.dev/npm/package/@react-ng/bounds-observer
> Additionally, this callback will be invoked on enabling, i.e. when the enabled property changes to true

ewanmellor replied:
@s77rt OK, I finally see what you are saying. I have updated this PR with all the changes mentioned.

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -13,14 +13,49 @@ import useWindowDimensions from '../../hooks/useWindowDimensions';
+ * enough then we go with that.
+ * @return {DOMRect} The chosen bounding box.
+ */
+function chooseBoundingBox(target, clientX, clientY, slop = 0) {
```
I think we'd need a safer algo without the recursive call.

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -13,14 +13,49 @@ import useWindowDimensions from '../../hooks/useWindowDimensions';
+        // Retry with a slop factor, in case the user is moving the mouse quickly.
+        return chooseBoundingBox(target, clientX, clientY, 5);
+    }
+    // Fall back to the full bounding box if we failed to find a matching one
+    // (shouldn't happen).
+    return target.getBoundingClientRect();
```
Don't write code that can't be reached.

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -13,14 +13,49 @@ import useWindowDimensions from '../../hooks/useWindowDimensions';
+ * @return {DOMRect} The chosen bounding box.
+ */
+function chooseBoundingBox(target, clientX, clientY, slop = 0) {
+    const bbs = target.getClientRects();
```
minor optimisation, if bbs contains only one element then just use it (that element would be same as getBoundingClientRect)

ewanmellor committed:
Fix: Chat - Link in end of line displays tooltip over text and not on link

Changes from PR review:
  o  Remove the shouldUseMultilinePositioning prop, and instead make the
new algorithm the one to use in all situations.
  o  Refactor chooseBoundingBox to avoid a recursive call.
  o  Shortcut chooseBoundingBox in the case where the element only
has one bounding box.
  o  Revert the changes to Hoverable that pass the MouseEvent through
onHoverIn / onHoverOut, and instead add direct pass-throughs for the
onMouseEnter / onMouseLeave events.
  o  Change the algorithm so that the onHoverIn event handler only
records the mouse position, and rely on BoundsObserver being enabled
as a side-effect of the isVisible change to move the tooltip.

ewanmellor commented:
recheck

s77rt commented on src/components/Hoverable/index.js:
```diff
@@ -159,13 +159,19 @@ class Hoverable extends Component {
                 if (_.isFunction(child.props.onMouseEnter)) {
                     child.props.onMouseEnter(el);
                 }
+                if (_.isFunction(this.props.onMouseEnter)) {
+                    this.props.onMouseEnter(el);
+                }
```
Move this to be the first callback (same for onMouseLeave)

ewanmellor replied:
Done.

s77rt replied:
This is still not the first, it needs to goes before setIsHovered

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -91,10 +156,15 @@ function Tooltip(props) {
-        setXOffset(bounds.x);
-        setYOffset(bounds.y);
+        const t = target.current;
+        if (!t) {
+            return;
+        }
```
Is there a case where target is not defined?

ewanmellor replied:
No, but the type system would not be able to see that, because it cannot tell that `updateBounds` will never be called before a hover event makes the tooltip visible.

Regardless, this file is not TypeScript, so I've removed it.

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -91,10 +156,15 @@ function Tooltip(props) {
+        if (!t) {
+            return;
+        }
+        const betterBounds = chooseBoundingBox(t, initialMousePosition.current.x, initialMousePosition.current.y);
```
Please add a comment here on why we need better bounding box

ewanmellor replied:
Done.

ewanmellor committed:
Fix: Chat - Link in end of line displays tooltip over text and not on link

Changes from PR review:
  o  Reorder the callbacks from Hoverable.
  o  Remove a null check.
  o  Add a comment.

ewanmellor committed:
Fix: Chat - Link in end of line displays tooltip over text and not on link

Changes from PR review:
  o  Small tidyup.

ewanmellor commented:
recheck

ewanmellor committed:
Fix: Chat - Link in end of line displays tooltip over text and not on link

Change from PR review: move the onMouseEnter / onMouseLeave callbacks
before the setIsHovered calls.

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -13,6 +13,63 @@ import useWindowDimensions from '../../hooks/useWindowDimensions';
+ * @param {Element} bbs The bounding boxes of DOM element being hovered over.
+ * @param {number} clientX The X position from the MouseEvent.
+ * @param {number} clientY The Y position from the MouseEvent.
+ * @param {number} slop An allowed slop factor when searching for the bounding
+ * box. If the user is moving the mouse quickly we can end up getting a
+ * hover event with the position outside any of our bounding boxes. We retry
+ * with a small slop factor in that case, so if we have a bounding box close
+ * enough then we go with that.
```
I have been checking and I think the need for the tolerance is not related to moving the mouse quickly, I can reproduce the bug constantly and this seems due to the returned bbs values corresponding to the text fragments only and not the block. Let's have the tolerance always used and integrate this function inside `chooseBoundingBox`

s77rt commented on src/components/Tooltip/index.js:
```diff
@@ -13,6 +13,63 @@ import useWindowDimensions from '../../hooks/useWindowDimensions';
+    // Fall back to the full bounding box if we failed to find a matching one.
+    // This could only happen if the user is moving the mouse very quickly
+    // and they got it outside our slop above.
+    return target.getBoundingClientRect();
```
```suggestion
    return bbs[0];
```
